exasol_transformers_extension/udfs/models/base_model_udf.py

```diff
@@ -0,0 +1,196 @@
+                self.clear_device_memory()
+                self.load_models(model_name)
+                self.last_loaded_model_key = current_model_key
+
+            yield model_df
+
+    def extract_unique_param_based_dataframes(
+            self, model_df: pd.DataFrame) -> Iterator[pd.DataFrame]:
+        """
+        Extract unique dataframes having same model parameter values. if there
+        is no model specified parameter, the input dataframe return as it is.
+
+        :param model_df: Dataframe used in prediction
+
+        :return: Unique model dataframes having specified parameters
+        """
+
+        yield model_df
+
+    def set_cache_dir(self, model_df: pd.DataFrame) -> None:
+        """
+        Set the cache directory in bucketfs of the specified model.
+
+        :param model_df: The model dataframe to set the cache directory
+        """
+        model_name = model_df['model_name'].iloc[0]
+        bucketfs_conn_name = model_df['bucketfs_conn'].iloc[0]
+        sub_dir = model_df['sub_dir'].iloc[0]
+        bucketfs_location = bucketfs_operations.create_bucketfs_location(
+            self.exa.get_connection(bucketfs_conn_name))
+
+        model_path = bucketfs_operations.get_model_path(sub_dir, model_name)
+        self.cache_dir = bucketfs_operations.get_local_bucketfs_path(
+            bucketfs_location=bucketfs_location, model_path=str(model_path))
+
+    def clear_device_memory(self):
+        """
+        Delete models and free device memory
+        """
+
+        del self.last_loaded_model
+        del self.last_loaded_tokenizer
+        torch.cuda.empty_cache()
+
+    def load_models(self, model_name: str, **kwargs) -> None:
+        """
+        Load model and tokenizer model from the cached location in bucketfs
+
+        :param model_name: The model name to be loaded
+        """
+        self.last_loaded_model = self.base_model.from_pretrained(
+            model_name, cache_dir=self.cache_dir)
+        self.last_loaded_tokenizer = self.tokenizer.from_pretrained(
+            model_name, cache_dir=self.cache_dir)
+        self.last_created_pipeline = self.pipeline(
+            self.task_name,
+            model=self.last_loaded_model,
+            tokenizer=self.last_loaded_tokenizer,
+            device=self.device,
+            framework="pt")
+
+    def get_prediction(self, model_df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform prediction of the given model and preparation of the prediction
+        results according to the format that the UDF can emit.
+
+        :param model_df: The dataframe to be predicted
+
+        :return: The dataframe where the model_df is formatted with the
+        prediction results
+        """
+        pred_df_list = self.execute_prediction(model_df)
+        pred_df = self.append_predictions_to_input_dataframe(
+            model_df, pred_df_list)
+        return pred_df
+
+    @staticmethod
+    def create_dataframes_from_predictions(
```

Comment:
Looks like this function is not used in the base class, if it is used in the subclass extract it as a utility function

Reply:
but, if it is not used in the abstract class, than we don't need to define it as abstract.

Reply:
It is necessary to create dataframes according to each model result. Is it better to put each one separately into utility or leave it in its own class?